### PR TITLE
Kernel class file and adaptation of instantaneous_rate function

### PR DESCRIPTION
**Pull request by dholstein**

### CODE REVIEW

File: elephant/kernels.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+import scipy.special
+
+default_kernel_area_fraction = 0.99999
+
+
+class Kernel(object):
+    """
+    Base class for kernels.
+
+    :param sigma: Standard deviation of the kernel.
```
Comment (alperyeg) [inline elephant/kernels.py]: Please adapt to numpy doc. 
```
Parameters
---------------
Sigma: ....
```

File: elephant/kernels.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+    having any effect in the case of symmetric kernels.
+    """
+    min_cutoff = np.sqrt(5.0)
+
+    def _evaluate(self, t):
+        return (3.0 / (4.0 * np.sqrt(5.0) * self._sigma_scaled)) * np.maximum(
+            0.0,
+            1 - (t / (np.sqrt(5.0) * self._sigma_scaled)).magnitude ** 2)
+
+    def boundary_enclosing_at_least(self,
```
Comment (alperyeg) [inline elephant/kernels.py]: Needs documentation.

File: elephant/kernels.py [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+    parameter `direction` needed for asymmetric kernels also exists without
+    having any effect in the case of symmetric kernels.
+    """
+    min_cutoff = 3.0
+
+    def _evaluate(self, t):
+        return (1 / (np.sqrt(2.0) * self._sigma_scaled)) * np.exp(
+            -(np.absolute(t) * np.sqrt(2.0) / self._sigma_scaled).magnitude)
+
+    def boundary_enclosing_at_least(self,
```
Comment (alperyeg) [inline elephant/kernels.py]: Here also documentation needed.

File: elephant/kernels.py [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+    def _evaluate(self, t):
+        """ Evaluates the kernel.
+
+        :param t: Interval on which the kernel is evaluated, not necessarily
+            a time interval.
+        :type t: Quantity 1D
+        :returns: The result of the kernel evaluations.
+        :rtype: Quantity 1D
+        """
+        raise NotImplementedError()
```
Comment (alperyeg) [inline elephant/kernels.py]: Do you intend to provide an implementation later, or why do you raise a  `NotImplementedError`?

File: elephant/kernels.py [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+        The formula in this method using retrieval of the sampling interval
+        from t only works for t with equidistant intervals!
+        The formula calculates the Median slightly wrong by the potentially
+        ignored probability in the distribution corresponding to lower values
+        than the minimum in the array t.
+        """
+        return np.nonzero(self(t).cumsum() *
+                          (t[len(t)-1] - t[0]) / (len(t) - 1) >= 0.5)[0].min()
+
+    def is_symmetric(self):
```
Comment (alperyeg) [inline elephant/kernels.py]: Shouldn't this return `True`?

File: elephant/kernels.py [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+        """ Should return `True` if the kernel is symmetric. """
+        return False
```
Comment (alperyeg) [inline elephant/kernels.py]: Shouldn't this return `True`?

File: elephant/statistics.py [hunk 7]
```diff
@@ -422,7 +433,8 @@ def make_kernel(form, sigma, sampling_period, direction=1):
     return kernel, norm, m_idx
 
 
-def instantaneous_rate(spiketrain, sampling_period, form,
+# to finally be taken out of Elephant
+def oldfct_instantaneous_rate(spiketrain, sampling_period, form,
```
Comment (alperyeg) [inline elephant/statistics.py]: Maybe a warning indicating that this function is deprecated  would be good.

File: elephant/kernels.py [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+
+
+class SymmetricKernel(Kernel):
+    """ Base class for symmetric kernels. """
+
+    # def __init__(self, sigma, direction):
```
Comment (alperyeg) [inline elephant/kernels.py]: Comments here should be removed.

**Comment by alperyeg** [comment]
Are the unittests for the kernel class the same as the old one? Was just wondering because I could not see any test file.

**Comment by dholstein** [comment]
test_statistics.py now adapted to the new version of the function instantaneous_rate and test_kernels.py added

File: elephant/kernels.py (hunk 1 above)
Comment (dholstein) [inline elephant/kernels.py]: Changed towards style of numpy doc.

File: elephant/kernels.py (hunk 4 above)
Comment (dholstein) [inline elephant/kernels.py]: The NotImplementedError message is now verbose. It is not intended to provide an implementation later, however, it is preferred to show the full list of existing methods of derived classes in the parent class for giving the user directly a list of methods to care for, if intending to supply further kernels in the future.

File: elephant/kernels.py (hunk 6 above)
Comment (dholstein) [inline elephant/kernels.py]: I think returning 'False' is correct here. In the case of symmetric kernels, this method is overwritten in the class SymmetricKernel and then returns 'True', remaining with 'False' for the asymmetric ones.

File: elephant/statistics.py (hunk 7 above)
Comment (dholstein) [inline elephant/statistics.py]: Several deprecation warnings introduced in statistics.py.

File: elephant/kernels.py (hunk 6 above)
Comment (alperyeg) [inline elephant/kernels.py]: Ok can you add your comment to the documentation? Then it should be clear.

File: elephant/kernels.py (hunk 2 above)
Comment (dholstein) [inline elephant/kernels.py]: Decorator inherit_docstring introduced, from which the methods boundary_enclosing_area_fraction and _evaluate now have documentation inherited from the parent class and optionally further kernel-specific documentation.

File: elephant/kernels.py (hunk 6 above)
Comment (dholstein) [inline elephant/kernels.py]: True, the documenting line wasn't very suitable. Changed now.

File: elephant/kernels.py (hunk 8 above)
Comment (dholstein) [inline elephant/kernels.py]: Comments of this type are removed now.

**Comment by dholstein** [comment]
Pullrequests for kernels.py and rate_estimation in statistics.py will be separated.

**Pull request closed without merging**